modin/core/execution/utils.py

```diff
@@ -29,3 +31,62 @@ def set_env(**environ):
     finally:
         os.environ.clear()
         os.environ.update(old_environ)
+
+
+if "_MODIN_DOC_CHECKER_" in os.environ:
+
+    # The doc checker should get the non-processed functions
+    def remote_function(func, ignore_defaults=False):
+        return func
+
+
+# Check if the function already exists to avoid circular imports
+elif "remote_function" not in dir():
+    from modin.config import Engine
+
+    if Engine.get() == "Ray":
+        from modin.core.execution.ray.common import RayWrapper
+
+        _preprocess_func = RayWrapper.put
+    elif Engine.get() == "Unidist":
+        from modin.core.execution.unidist.common import UnidistWrapper
+
+        _preprocess_func = UnidistWrapper.put
+    elif Engine.get() == "Dask":
+        from modin.core.execution.dask.common import DaskWrapper
+
+        # The function cache is not supported for Dask
+        def remote_function(func, ignore_defaults=False):
+            return DaskWrapper.put(func)
+
+    else:
+
+        def remote_function(func, ignore_defaults=False):
+            return func
+
+    if "remote_function" not in dir():
+        _remote_function_cache = {}
+
+        def remote_function(func, ignore_defaults=False):  # noqa: F811
+            if "<locals>" in func.__qualname__:  # Nested function
```

Comment:
I think we should avoid all functions with closure, not only local ones. If a global function uses a global variable, then the value from closure will be placed in an object storage only once, if the global value is changed it will be ignored:

<details><summary>example</summary>

```python
import ray

_cache = {}

def remote_function(func, ignore_defaults=False):
    if "<locals>" in func.__qualname__:  # Nested function
        if func.__closure__:
            return func
        if not ignore_defaults and func.__defaults__:
            return func
        # For the nested functions, use __code__ as the key
        key = id(func.__code__)
    else:
        key = func
    ref = _cache.get(key, None)
    if ref is None:
        ref = ray.put(func)
        _cache[key] = ref
    return ref

@ray.remote
def func(fn):
    fn()

global_var = "first func"

@remote_function
def global_func():
    print(global_var)

ray.get(func.remote(global_func))

global_var = "second func"

ray.get(func.remote(global_func))

# Output:
# first func
# first func
```

</details>

Reply:
Also a question to @YarShev as you're more familiar with this stuff: could there be any problems with caching if a cached function uses some module from a global scope? Like:
```python
import numpy as np

@remote_function # caching
def fn(x):
    return np.exp(x)
```

Reply:
There could be a problem if fn is a remote function on a worker and the worker doesn't have the module loaded in it.

Reply:
Are you sure a global function may have a closure?
```python
>>> A='b'
>>> def f():
...   global A
...   A='c'
...   print(A)
... 
>>> f()
c
>>> print(f.__closure__)
None
>>> def f():
...   print(A)
... 
>>> f()
c
>>> print(f.__closure__)
None
```

Reply:
> could there be any problems with caching if a cached function uses some module from a global scope?

What's the difference between cached and non-cached function in this case?